solarkraft/src/state/value.ts

```diff
@@ -16,7 +16,98 @@ export type Value = {
     | MapValue
 )
 
-export function isValid(v: Value): boolean {
+const EMPTY_COLLECTION_TYPE: string = 'T'
+const UNKNOWN_TYPE: string = 'UNKNOWN'
+
+// Helper function, takes an array of strings (expected: types of child values for a Vec or Map Value)
+// and returns:
+//  - an empty array, iff the input array is empty
+//  - an array with a single element, iff every element within the input array is equal to that element
+//  - the original array, if two or more elements in the input array are different
+// whenever the function returns a singleton array, we can consider the original collection to be homogeneous
+// and its contents to be of the type contained in the singleton return.
+function computeSingleTypeRep(ts: string[]): string[] {
+    if (ts.length === 0) return []
+    else {
+        const t0 = ts[0]
+        const hom = !ts.some((t) => t !== t0)
+        if (hom) return [t0]
+        else return ts
+    }
+}
+
+// Returns the full type annotation of the provided (possibly complex) Value `v`.
+// It uses Apalache-style type notation, with Soroban type primitives.
```

Comment:
what is Apalache-style type notation? What we see below looks like Soroban to me.

Reply:
Well, our "maps" are `A -> B`